src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java

```diff
@@ -0,0 +1,106 @@
+import core.basesyntax.model.User;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+class RegistrationServiceImplTest {
+    private static RegistrationService registrationService;
+    private static StorageDao storageDao;
+    private static final String INVALID_LOGIN = "user";
+    private static final String VALID_LOGIN = "user1986";
+    private static final String INVALID_PASSWORD = "pass";
+    private static final String VALID_PASSWORD = "password";
+    private static final String EXIST_LOGIN = "bob1989";
+    private static final int INVALID_AGE = 15;
+    private static final int VALID_AGE = 25;
+    private User user;
+
+    @BeforeAll
+    static void beforeAll() {
+        registrationService = new RegistrationServiceImpl();
+        storageDao = new StorageDaoImpl();
+        User bob = new User();
+        bob.setLogin(EXIST_LOGIN);
+        storageDao.add(bob);
+    }
+
+    @BeforeEach
+    void setUp() {
+        user = new User();
+    }
+
+    @Test
+    void register_nullUser_notOk() {
+        try {
+            registrationService.register(null);
+        } catch (RegistrationException e) {
+            return;
+        }
+        fail("If user is null RegistrationException should be.");
+    }
+
+    @Test
+    void register_nullLogin_NotOk() {
+        user.setLogin(null);
+        assertThrows(RegistrationException.class, () -> registrationService.register(user));
+    }
+
+    @Test
+    void register_nullAge_notOk() {
+        user.setAge(null);
+        assertThrows(RegistrationException.class, () -> registrationService.register(user));
+
+    }
+
+    @Test
+    void register_nullPassword_notOk() {
+        user.setPassword(null);
+        assertThrows(RegistrationException.class, () -> registrationService.register(user));
+
+    }
+
+    @Test
+    void register_ageLess18_notOk() {
+        user.setAge(INVALID_AGE);
+        assertThrows(RegistrationException.class, () -> registrationService.register(user));
+
+    }
+
+    @Test
+    void register_PasswordLengthLess6Chars_notOk() {
+        user.setPassword(INVALID_PASSWORD);
+        assertThrows(RegistrationException.class, () -> registrationService.register(user));
+    }
+
+    @Test
+    void register_LoginLengthLess6Chars_notOk() {
+        user.setLogin(INVALID_LOGIN);
+        assertThrows(RegistrationException.class, () -> registrationService.register(user));
+
+    }
+
+    @Test
+    void register_ExistUser_notOk() {
+        user.setLogin(EXIST_LOGIN);
+        assertThrows(RegistrationException.class, () -> registrationService.register(user));
+    }
+
+    @Test
+    void register_UserIsValid() {
+        user.setLogin(VALID_LOGIN);
+        user.setAge(VALID_AGE);
+        user.setPassword(VALID_PASSWORD);
+        assertDoesNotThrow(() -> registrationService.register(user),
+                "Registering user with valid data should not throw an exception");
+
+    }
+}

```

Comment:
in this tests we should arrenage the conditions before saving the user, by that I mean that we should put some user in the `Storage`(not using the `register()` method, check common mistakes) and after that try to register another user with the same login and assert an exception that we get from that.

BTW this logic that you have here now, should not throw any exception from the looks of it. If you want we can also investigate this part and try to figure out why it does

Reply:
btw, @rel1c-hub if you notice that something is wrong, next time try to explain why you think it is, and to fix that together first :wink:

Reply:
this test is passing not because of the reasons we think it is, but because it has a null password and age, this is why need to assert the exception message - because it saves us from cases like that, when we think that it passes and all Is great, when in the reality it is written wrong and we wouldn't know if this functionality is broken.
Try to strictly follow my instructions that I have wrote in the previous comment about it,
here is another pseudo-code example:
```java
 @Test
    void register_userWithTheSameLogin_notOk() {
        // create some user or use the one that we already have
        // put it into storage via dao class or storage directly, because we cannot use register for it due to reason described above and in common mistakes
        // create another user wit the same login
        // register that second user
        // assert the resulting error
        // assert the message
}
```